src/js/components/Button/Badge.js

```diff
@@ -0,0 +1,161 @@
+import React, { useContext, useEffect, useRef, useState } from 'react';
+import { ThemeContext } from 'styled-components';
+import { parseMetricToNum } from '../../utils';
+import { Box } from '../Box';
+import { Stack } from '../Stack';
+import { Text } from '../Text';
+
+const getBadgeDimension = (dimension, content, badgeContentRef, theme) => {
+  if (
+    typeof content === 'number' ||
+    (typeof content === 'object' && content.value)
+  ) {
+    const borderWidth = content.border
+      ? parseMetricToNum(theme.global.borderSize[content.border.size]) * 2
+      : 0;
+    // leave space to pad content, must be incorporated in width as opposed
+    // to conventional pad on Box
+    let { pad } = theme.button.badge.container;
+    if (typeof pad === 'string')
+      pad = parseMetricToNum(theme.global.edgeSize[pad] || pad);
+    else if (dimension === 'height') {
+      if (pad.top)
+        pad = parseMetricToNum(theme.global.edgeSize[pad.top] || pad.top);
+      if (pad.bottom)
+        pad += parseMetricToNum(
+          theme.global.edgeSize[pad.bottom] || pad.bottom,
+        );
+      else if (pad.vertical) {
+        pad =
+          parseMetricToNum(
+            theme.global.edgeSize[pad.vertical] || pad.vertical,
+          ) * 2;
+      } else pad = 0;
+    } else if (dimension === 'width') {
+      if (pad.left)
+        pad = parseMetricToNum(theme.global.edgeSize[pad.left] || pad.left);
+      if (pad.right)
+        pad += parseMetricToNum(theme.global.edgeSize[pad.right] || pad.right);
+      else if (pad.horizontal) {
+        pad =
+          parseMetricToNum(
+            theme.global.edgeSize[pad.horizontal] || pad.horizontal,
+          ) * 2;
+      } else pad = 0;
+    }
+
+    // if content is tall/wide, let badge grow to fit. otherwise,
+    // make sure it's at least badge.size.medium dimensions
+    return `${Math.max(
```

Comment:
This function feels like we're doing stuff the browser is already doing. Would it be possible to just set `min-width` and then adjust the offset via `useLayoutEffect()` where we can measure how big it actually is?